---
title: "Theo's version"
author: "Theo-60985751, Hannah- , Jingxuan-"
date: "2023-11-18"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Q1. Sequencing technologies

Areas of the genome with high GC content are harder to sequence because these 
regions coil up to themselves and require energy to separate them into straight
strands that can finally be PCR-ed/sequenced. 
Source: https://www.neb.com/en/nebinspired-blog/four-tips-for-pcr-amplification
-of-gc-rich-sequences#:~:text=Why%20can%20these%20regions%20be,break%20the%20th
ree%20hydrogen%20bonds.

# Q2. Global alignment exercise

# Q3. Looking at the Metadata of an alignment (SAM) file

## Q3.1

```{r}
data_metadat_sam = read.csv("single_cell_RNA_seq_bam.sam", nrows=73, sep="\t", header=FALSE,
fill=TRUE)
```

SN is reference sequence name, and LN is reference sequence length with range
[1, 2^31 âˆ’ 1]

## Q3.2 

```{r}
cat("Length of our X chromosome aligntment: ", data_metadat_sam[22,3])
```


# Q4. Looking at the Reads of an alignment (SAM) file

## Q4.1 

```{r}
sam <- read.csv("single_cell_RNA_seq_bam.sam", sep="\t", header=FALSE,
comment.char="@", col.names = paste0("V",seq_len(30)), fill=TRUE)
sam <- sam[paste0("V",seq_len(11))]

cat("Number of reads in this BAM file:", nrow(sam))
```

## Q4.2 

```{r}
print(sam[10,])
```
We should look at column "V3" to find the chromosome to which the read was 
aligned (the reference sequence). Column "V11" is the quality score in ASCII
base. 

## Q4.3

```{r}
sam$X_allign = sam$V3 == 'X'
cat("There are", sum(sam$X_allign), "reads that alligns to chromosome X")

```


## Q4.4

```{r}
phred33toQ = function(ascii){
  return( as.numeric(charToRaw(ascii)) - 33)
}

sam_xchromo = sam[which(sam$V3 == "X"),]


total_mean_quality_xchromo = 0
for (quality in sam_xchromo$V11){
  # print(mean(phred33toQ(quality)))
  total_mean_quality_xchromo = total_mean_quality_xchromo + mean(phred33toQ(quality))
}

mean_quality_xchromo = total_mean_quality_xchromo/nrow(sam_xchromo)

cat("The mean base quality for reads aligning to chromosome X is", mean_quality_xchromo)
```

## Q4.5

including libraries
```{r}
library(ggplot2)
library(tidyverse)
library(patchwork)
```

separating the sam file into smaller ones

mini_sam1
```{r}

mini_sam1 = sam[1:50000,]

df_quality_1 = matrix(nrow = 58, ncol = 1)
read_number = 1
for (read in mini_sam1$V11){
  quality = phred33toQ(read)
  col_name = paste0("read", read_number)
  df_quality_1 = cbind(df_quality_1, quality)
  read_number = read_number + 1
}

df_quality_1 = df_quality_1[,-c(1)]
# df_quality_1 = t(df_quality_1)
# colnames(df_quality_1) = c(1:58)
# df_quality_1 = stack(as.data.frame(df_quality_1))
# 
# ggplot(data = df_quality_1)+
#   geom_boxplot(aes(x=ind, y = values))

```


mini_sam2

```{r}
mini_sam2 = sam[50001:100000,]

df_quality_2 = matrix(nrow = 58, ncol = 1)
read_number = 1001
for (read in mini_sam2$V11){
  quality = phred33toQ(read)
  col_name = paste0("read", read_number)
  df_quality_2 = cbind(df_quality_2, quality)
  read_number = read_number + 1
}
df_quality_2 = df_quality_2[,-c(1)]



```

mini_sam3
```{r}
mini_sam3 = sam[100001:146346,]

df_quality_3 = matrix(nrow = 58, ncol = 1)
read_number = 2001
for (read in mini_sam3$V11){
  quality = phred33toQ(read)
  col_name = paste0("read", read_number)
  df_quality_3 = cbind(df_quality_3, quality)
  read_number = read_number + 1
}
df_quality_3 = df_quality_3[,-c(1)]



```


combining smaller sam quality scores into one big file
```{r}

df_quality = cbind(df_quality_1, df_quality_2)
df_quality = cbind(df_quality, df_quality_3)

df_quality = t(df_quality)
colnames(df_quality) = c(1:58)
df_quality = stack(as.data.frame(df_quality))

```


plotting the barplot
```{r}
ggplot(data = df_quality)+
  geom_boxplot(aes(x=ind, y = values))

```


## Q4.6

Column 4 contains the leftmost mapping position of the read

## Q4.7

```{r}
sam$hspa8_allign = sam$V4 > 40801273 & sam$V4 < 40805199

cat("There are", sum(sam$hspa8_allign), "reads that have their leftmost mapping
    position aligned with the cordinate for Hspa8 protein")
```


## Q4.8

```{r}
sam$MAPQ_50less = sam$V5 < 50

cat("There are",sum(sam$MAPQ_50less), "reads that have mapping quality score less than 50")
```

## Q4.9

```{r}

```


