---
title: "Theo's version"
author: "Theo-60985751, Hannah- , Jingxuan-"
date: "2023-11-18"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Q1. Sequencing technologies

Areas of the genome with high GC content are harder to sequence because these 
regions coil up to themselves and require energy to separate them into straight
strands that can finally be PCR-ed/sequenced. 
Source: https://www.neb.com/en/nebinspired-blog/four-tips-for-pcr-amplification
-of-gc-rich-sequences#:~:text=Why%20can%20these%20regions%20be,break%20the%20th
ree%20hydrogen%20bonds.

# Q2. Global alignment exercise

# Q3. Looking at the Metadata of an alignment (SAM) file

## Q3.1

```{r}
data_metadat_sam = read.csv("single_cell_RNA_seq_bam.sam", nrows=73, sep="\t", header=FALSE,
fill=TRUE)
```

SN is reference sequence name, and LN is reference sequence length with range
[1, 2^31 âˆ’ 1]

## Q3.2 

The length of the X chromosome for our alignment is the LN value 

# Q4. Looking at the Reads of an alignment (SAM) file

## Q4.1 

```{r}
sam <- read.csv("single_cell_RNA_seq_bam.sam", sep="\t", header=FALSE,
comment.char="@", col.names = paste0("V",seq_len(30)), fill=TRUE)
sam <- sam[paste0("V",seq_len(11))]

concat("Number of reads in this BAM file is:", length(sam)
```


